mongo/src/main/java/com/redhat/lightblue/mongo/crud/BsonMerge.java

```diff
@@ -146,6 +148,62 @@ protected int size(List value) {
         return value.size();
     }
 
+    @Override
+    public IdentityExtractor getArrayIdentityExtractor(Path arrayField) {
+        Path array=getArray(arrayField);
+        if(ignoredIdentities.contains(array))
+            return null;
+        else {
+            ArrayIdentityFields fields = getArrayIdentities().get(array);
+            if (fields != null) {
+                return getArrayIdentityExtractorImpl(fields);
+            }
+        }
+        return null;
+    }
+    
+        /**
+     * Input is a path to an array element
+     * Output is the array containing that element, and all indexes replaced with *
+     */
+    private Path getArray(Path p) {
+        int lastIndex=p.numSegments()-1;
+        while(lastIndex>0) {
+            if(p.isIndex(lastIndex))
+                break;
+            lastIndex--;
+        }
+        MutablePath mp=new MutablePath();
+        for(int i=0;i<lastIndex;i++) {
+            if(p.isIndex(i))
+                mp.push(Path.ANY);
+            else
+                mp.push(p.head(i));
+        }
+        return mp.immutableCopy();
+    }
+
+    public DocComparator.Difference<Object> compareNodesWithFallback(DBObject oldDoc,DBObject newDoc) throws Exception {
+        // Compare two docs. If we get a duplicate array identity exception, remove that array identity from
+        // the identity map, and compare again. This is to deal with docs that contain duplicate array identities either
+        // because the identity constraint is added after data is screwed up, or becase we forgot to validate it
+        while(true) {
+            try {
+                return compareNodes(oldDoc,newDoc);
+            } catch(DocComparator.DuplicateArrayIdentity dai) {
+                LOGGER.debug("Duplicate ID:{}");
```

Comment:
How about a warning (log.warn) about a non unique identity in an array? This is a data integrity problem after all.